### PR TITLE
use external rng instead of nimcrypto

### PR DESCRIPTION
**Pull request by arnetheduck**
allows passing in different RNG's to generate keys

### CODE REVIEW

**Comment by arnetheduck** [comment]
>  I would be more comfortable if we were completely deterministic and only passing an Input-Keying-Material

well, I suspect this would simply move the loop outside - ie there's not a lot of options the caller has if they want a key - if they care enough, they could pass in an RNG that simply fails after an iteration, but in the most common case, they just want key no matter what.

that said, I don't really mind changing this to take an array of random bytes - as far as secp is concerned, we just pass that through `verify` and we're done (as long as there's enough bytes). The API becomes slightly more tricky because now you need to verify the length of it, and you need tell the caller that it's "safe" to try again with a different number etc so there's a few more ways to fail when calling it.

> constant-time

Is this relevant in the context? ie that the RNG produces a quality number (presumably in constant time) seems important but how many tries we take to get there seems less of an issue?

> entropy

the nist-based CSPRNGS run out after something like 2^48 bytes - I think we can safely ignore this for anything but the most secure generation methods - ie for generating long-running keys like your main wallet, there exist better ways - this function is basically here for the default "let's set up an ephemeral secure channel" kind of use

File: tests/test_secp256k1.nim [hunk 1]
```diff
@@ -1,10 +1,22 @@
-import ../secp256k1, unittest
+import ../secp256k1, unittest, random
```
Comment (cheatfate) [inline tests/test_secp256k1.nim:1]: As you said many times tests are "examples". And if you are using insecure `random` package it means that somebody will start using it too. We should avoid import of `random` at any cost, your `unsafeRng` could generate static data, and produce static private key, but at least we will not show this "unsafe" approach.

Comment (mratsim) [reply tests/test_secp256k1.nim:1]: `random` can be imported but only in a `when isMainModule` section

**Comment by mratsim** [comment]
> well, I suspect this would simply move the loop outside - ie there's not a lot of options the caller has if they want a key

The scheme used in Eth2 always produce a valid key with a single RNG call. It works like this:
1. Get a random input
2. Take the input modulo r, the curve order

What we are doing here is just testing repeatedly until we find an integer below the curve order.

In the BLS spec, the Get a random input is split into
1. Get the output of a CSPRNG as an 32 bytes IKM
2. HKDF it

> > constant-time
>
> Is this relevant in the context? ie that the RNG produces a quality number (presumably in constant time) seems important but
> how many tries we take to get there seems less of an issue?

It is, source https://cfrg.github.io/draft-irtf-cfrg-hash-to-curve/draft-irtf-cfrg-hash-to-curve.html
image

The first iteration of WPA3 was broken due to non-constant-time password encoding:
- https://eprint.iacr.org/2019/383
- Writeup: https://wpa3.mathyvanhoef.com/

As long as we touch secret data, we SHOULD be constant-time.

File: secp256k1.nim [hunk 2]
```diff
@@ -165,11 +167,19 @@ func fromHex*(T: type seq[byte], s: string): SkResult[T] =
...
-proc random*(T: type SkSecretKey): SkResult[T] =
-  ## Generates new random private key.
+type
+  Rng* = proc(buf: var openArray[byte]): bool {.raises: [Defect].}
+    ## Function filling the given buffer with random data, or returning false
+
+proc random*(T: type SkSecretKey, rng: Rng): SkResult[T] =
+  ## Generates new random private key - a high-quality RNG should be used -
+  ## see nimcrypto or bearssl for good RNG's. The random number generator in
+  ## the Nim standard library `random` module in particular is of poor quality.
```
[the remark this answers is not in the document]
Comment (arnetheduck) [reply secp256k1.nim]: fixed

File: secp256k1.nim [hunk 3]
```diff
@@ -357,9 +367,9 @@ func toRaw*(sig: SkRecoverableSignature): array[SkRawRecoverableSignatureSize, b
 func toHex*(sig: SkRecoverableSignature): string =
   toHex(toRaw(sig))
 
-proc random*(T: type SkKeyPair): SkResult[T] =
+proc random*(T: type SkKeyPair, rng: Rng): SkResult[T] =
   ## Generates new random key pair.
-  let seckey = ? SkSecretKey.random()
+  let seckey = ? SkSecretKey.random(rng)
```
[the remark this answers is not in the document]
Comment (arnetheduck) [reply secp256k1.nim:369]: ok - leaving the point selection for another day, and skipping the callback for an API that may fail instead.

**Pull request merged**